### PR TITLE
Add Share button to the SelectableRegion toolbar on Android

### PR DESCRIPTION
**Pull request by bleroux**
## Description

This PR adds the share button to text selection toolbar buttons on Android ~~and iOS~~ for `SelectableRegion` (and therefore `SelectionArea`).

https://github.com/flutter/flutter/pull/139479 adds this button for `EditableText` (which is used by `TextField` and `SelectableText` but not by `SelectionArea`).

**Edit**: supporting this on iOS will need more work (see https://github.com/flutter/flutter/pull/141447#issuecomment-1889942622 and https://github.com/flutter/flutter/issues/141775).

## Related Issue

Follow up for https://github.com/flutter/flutter/issues/138728

## Tests

Adds 1 test.

### CODE REVIEW

**Comment by ueman** [comment]
Does this work on an iPad? The share functionality on the iPad needs an anchor for the share popup. For the textfield it works because it's backed on iOS with an invisible native text field, which I assume is not the case for text selection.

**Comment by bleroux** [comment]
Very good point @ueman :+1: 
@justinmc As a follow up to  https://github.com/flutter/flutter/issues/138728, do you think it would be wiser that I made this PR Android only?
The iOS will need more work, probably some on the engine side (similar to https://github.com/flutter/engine/pull/48427).

**Comment by ueman** [comment]
I think long term, the framework needs to pass the anchor to the engine. Otherwise, people will keep repeating that mistake since it's really not that obvious 😕

**Comment by bleroux** [comment]
@justinmc I have updated the PR to make it Android only and I filed https://github.com/flutter/flutter/issues/141775 to track work on the iOS side.

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 1]
```diff
@@ -267,9 +267,28 @@ class SelectableRegion extends StatefulWidget {
...
     final bool canSelectAll = selectionGeometry.hasContent;
+    final bool canShare = onShare != null && switch (defaultTargetPlatform) {
+      TargetPlatform.android
+        => selectionGeometry.status == SelectionStatus.uncollapsed,
+      TargetPlatform.macOS
+      || TargetPlatform.fuchsia
+      || TargetPlatform.linux
+      || TargetPlatform.windows
+        => false,
+      // TODO(bleroux): the share button should be proposed on iOS but the share
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/selectable_region.dart:282]: nit: consider `proposed` -> `shown`.

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 2]
```diff
@@ -267,9 +267,28 @@ class SelectableRegion extends StatefulWidget {
     required final SelectionGeometry selectionGeometry,
     required final VoidCallback onCopy,
     required final VoidCallback onSelectAll,
+    required final VoidCallback? onShare,
   }) {
     final bool canCopy = selectionGeometry.status == SelectionStatus.uncollapsed;
     final bool canSelectAll = selectionGeometry.hasContent;
+    final bool canShare = onShare != null && switch (defaultTargetPlatform) {
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/selectable_region.dart:274]: nit: consider making the switch its own variable maybe, `platformCanShare`.

File: packages/flutter/test/widgets/selectable_region_test.dart [hunk 3]
```diff
@@ -3321,14 +3332,73 @@ void main() {
...
+    buttonItems[1].onPressed?.call();
+    expect(lastShare, 'are');
+    // On Android, share should clear the selection.
+    expect(regionState.selectionOverlay, isNull);
+    expect(regionState.selectionOverlay?.startHandleLayerLink, isNull);
+    expect(regionState.selectionOverlay?.endHandleLayerLink, isNull);
   },
-    skip: kIsWeb, // [intended]
-    variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS, TargetPlatform.android, TargetPlatform.fuchsia }),
+    skip: kIsWeb, // [intended] Web uses its native context menu.
```
Comment (Renzo-Olivares) [inline packages/flutter/test/widgets/selectable_region_test.dart:3397]: should we only run this test on `Android`?

Comment (bleroux) [reply packages/flutter/test/widgets/selectable_region_test.dart:3397]: Yes this test is Android only on purpose. This is the new test I added.
The Github diff makes it look like I removed the variant line for it, but this is a variant line I updated for another test (see lines 3336-3337).

Testing if the share button is visible on other platforms is done in a test below (lines 3440-3455).
This new test checks the behavior when clicking on the share button (this is why it is Android only).

**Review by Renzo-Olivares — APPROVED** [review]
LGTM

**Pull request merged**